File: scripts/plot_benchmark.jl
using Plots
using DelimitedFiles
using Printf
using LaTeXStrings

# ==========================================
# 配置
# ==========================================
csv_file = "benchmark_beta_scan.csv"

# 自定义 X 轴刻度 (10^0 到 10^3)
my_xticks = [1, 10, 100, 1000]

# 全局画图样式设置
default(
    grid = true, 
    minorgrid = true,          # 显示次级网格 (对数坐标很需要)
    frame = :box,              # 方框围起来
    lw = 2.0,                  # 线宽
    markersize = 5,            # 散点大小
    xscale = :log10,           # X轴对数
    xticks = my_xticks,        # 强制显示 1, 10, 100, 1000
    guidefontsize = 12,        # 轴标签字体大小
    tickfontsize = 10,         # 刻度字体大小
    legendfontsize = 9,        # 图例字体大小
    margin = 5Plots.mm,         # 边距
    dpi = 300                   # 分辨率
)

# ==========================================
# 读取数据
# ==========================================
if !isfile(csv_file)
    println("Error: File '$csv_file' not found in current directory.")
    exit(1)
end

println("Reading data from $csv_file ...")

# 读取 CSV，跳过第一行表头
# data 是一个 Matrix{Any}，我们需要转换类型
raw_data, header = readdlm(csv_file, ',', header=true)

# 确保转换为 Float64
data = Float64.(raw_data)

# 根据之前的 header 顺序提取列:
# Beta, AccRate, Global, Err_Global, Pair, Err_Pair, RHS, Diff, Err_Diff
betas      = data[:, 1]
acc_rates  = data[:, 2]
avg_global = data[:, 3]
err_global = data[:, 4]
avg_pair   = data[:, 5]
err_pair   = data[:, 6]
avg_rhs    = data[:, 7]
avg_diff   = data[:, 8]
err_diff   = data[:, 9]

# ==========================================
# 画图 1: Order Parameters (Values)
# ==========================================
println("Plotting Order Parameters...")

p1 = plot(xlabel=L"$\beta$", ylabel=L"$|\Delta|$")

# 1. HMC Global
plot!(p1, betas, avg_global, yerror=err_global, 
      label=L"HMC: $|\Delta|$", marker=:circle, color=:blue)

# 2. HMC Pair (稍微错开一点点或者用不同的形状，防止重叠看不清)
plot!(p1, betas, avg_pair, yerror=err_pair, 
      label=L"HMC: $|\langle\hat{\Delta}\rangle|$", marker=:rect, color=:red, alpha=0.7)

# 3. BCS RHS (理论值)
plot!(p1, betas, avg_rhs, 
      label=L"BCS: $|\langle\hat{\Delta}\rangle|$", ls=:dash, color=:black, lw=2.5)

# 保存
savefig(p1, "plot_benchmark_values.png")


# ==========================================
# 画图 2: Consistency Check (Differences)
# ==========================================
println("Plotting Differences...")

# 计算差值
diff_GP = avg_global .- avg_pair
diff_GR = avg_global .- avg_rhs

# 误差传播: err = sqrt(err1^2 + err2^2)
# RHS 是解析解，假设无误差，所以 diff_GR 的误差就是 err_global
err_GP = sqrt.(err_global.^2 .+ err_pair.^2)

p2 = plot(xlabel="Inverse Temperature (β)", ylabel="Difference", 
          title="Consistency Check")

# 1. Global - Pair
plot!(p2, betas, diff_GP, yerror=err_GP, 
      label="Global - Pair", marker=:diamond, color=:purple)

# 2. Global - RHS
plot!(p2, betas, diff_GR, yerror=err_global, 
      label="Global - RHS", marker=:utriangle, color=:orange)

# 3. HMC Diff (Internal consistency)
plot!(p2, betas, avg_diff, yerror=err_diff, 
      label="HMC Δ_diff", marker=:hline, color=:green)

# 添加一条 y=0 的参考线
hline!(p2, [0.0], label="", color=:grey, ls=:dot)

# 保存
savefig(p2, "plot_benchmark_errors.png")


# ==========================================
# 画图 3 (额外): Acceptance Rate
# ==========================================
println("Plotting Acceptance Rates...")

p3 = plot(xlabel="Inverse Temperature (β)", ylabel="Acceptance Rate", 
          title="HMC Acceptance Rate", ylims=(0, 1.1))

plot!(p3, betas, acc_rates, 
      label="Acc Rate", marker=:star, color=:darkgreen, lw=1.5)

# 保存
savefig(p3, "plot_benchmark_acc_rate.png")

println("Done! Generated: plot_benchmark_values.png, plot_benchmark_errors.png, plot_benchmark_acc_rate.png")

File: scripts/plot_benchmark.py
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os

# ==========================================
# 配置
# ==========================================
csv_file = "benchmark_beta_scan.csv"
output_dpi = 300  # 输出图片分辨率

# 设置出版级绘图风格
plt.rcParams.update({
    # "font.family": "serif",          # 使用衬线字体 (类似 Times New Roman)
    # "font.serif": ["Times New Roman", "DejaVu Serif"],
    "font.size": 14,                 # 全局字号
    "axes.titlesize": 16,
    "axes.labelsize": 16,
    "xtick.labelsize": 14,
    "ytick.labelsize": 14,
    "legend.fontsize": 12,
    "xtick.direction": "in",         # 刻度朝内
    "ytick.direction": "in",
    "lines.linewidth": 2,
    "lines.markersize": 6,
    "figure.figsize": (8, 6),        # 图片默认大小
    "mathtext.fontset": "cm"         # 数学公式使用 Computer Modern 字体
})

# ==========================================
# 读取数据
# ==========================================
if not os.path.exists(csv_file):
    print(f"Error: {csv_file} not found.")
    exit(1)

print(f"Reading data from {csv_file}...")
df = pd.read_csv(csv_file)

# 数据列名参考:
# Beta,AccRate,Global,Err_Global,Pair,Err_Pair,RHS,Diff,Err_Diff

# 计算额外的差值列
# Diff_GP = Global - Pair
# Diff_GR = Global - RHS
# 误差传播: sqrt(err1^2 + err2^2)
diff_gp = df['Global'] - df['Pair']
diff_gr = df['Global'] - df['RHS']
err_gp = np.sqrt(df['Err_Global']**2 + df['Err_Pair']**2)

# 定义自定义 X 轴刻度
custom_xticks = [1, 10, 100, 1000]
custom_xticklabels = [r"$10^0$", r"$10^1$", r"$10^2$", r"$10^3$"]

# ==========================================
# 图 1: Order Parameters (Values)
# ==========================================
print("Plotting Order Parameters...")
fig, ax = plt.subplots()

# 1. HMC Global
ax.errorbar(df['Beta'], df['Global'], yerr=df['Err_Global'], 
            fmt='o', capsize=3, label='HMC Global', color='tab:blue', zorder=3)

# 2. HMC Pair (稍微错开一点点 x 轴，防止重叠)
ax.errorbar(df['Beta'] * 1.05, df['Pair'], yerr=df['Err_Pair'], 
            fmt='s', capsize=3, label='HMC Pair', color='tab:red', alpha=0.8, zorder=2)

# 3. BCS RHS (理论值 - 线)
ax.plot(df['Beta'], df['RHS'], linestyle='--', color='black', label='BCS RHS', zorder=1)

# 设置坐标轴
ax.set_xscale('log')
ax.set_xlabel(r'Inverse Temperature $\beta$')
ax.set_ylabel(r'Order Parameter $|\Delta|$')
ax.set_title('Order Parameter Benchmark (Clean Limit)')

# 设置刻度
ax.set_xticks(custom_xticks)
ax.set_xticklabels(custom_xticklabels)
ax.grid(True, which="both", ls="-", alpha=0.2) # 网格

ax.legend(loc='lower right', frameon=True, framealpha=0.9)

plt.tight_layout()
plt.savefig('py_benchmark_values.png', dpi=output_dpi)
plt.close()

# ==========================================
# 图 2: Consistency Check (Differences)
# ==========================================
print("Plotting Differences...")
fig, ax = plt.subplots()

# 参考线 y=0
ax.axhline(0, color='gray', linestyle=':', linewidth=1.5)

# 1. Global - Pair
ax.errorbar(df['Beta'], diff_gp, yerr=err_gp, 
            fmt='D', capsize=3, label=r'Global $-$ Pair', color='tab:purple')

# 2. Global - RHS
ax.errorbar(df['Beta'], diff_gr, yerr=df['Err_Global'], 
            fmt='^', capsize=3, label=r'Global $-$ RHS', color='tab:orange')

# 3. HMC Diff (Internal)
ax.errorbar(df['Beta'], df['Diff'], yerr=df['Err_Diff'], 
            fmt='_', capsize=3, label=r'HMC $\Delta_{\mathrm{diff}}$', color='tab:green', markersize=10)

# 设置坐标轴
ax.set_xscale('log')
ax.set_xlabel(r'Inverse Temperature $\beta$')
ax.set_ylabel('Difference')
ax.set_title('Consistency Check')

# 设置刻度
ax.set_xticks(custom_xticks)
ax.set_xticklabels(custom_xticklabels)
ax.grid(True, which="both", ls="-", alpha=0.2)

ax.legend(loc='best', frameon=True, framealpha=0.9)

plt.tight_layout()
plt.savefig('py_benchmark_errors.png', dpi=output_dpi)
plt.close()

# ==========================================
# 图 3: Acceptance Rate
# ==========================================
print("Plotting Acceptance Rates...")
fig, ax = plt.subplots()

ax.plot(df['Beta'], df['AccRate'], marker='*', linestyle='-', color='tab:green', markersize=8)

ax.set_xscale('log')
ax.set_xlabel(r'Inverse Temperature $\beta$')
ax.set_ylabel('Acceptance Rate')
ax.set_title('HMC Acceptance Rate')
ax.set_ylim(-0.05, 1.05)

# 设置刻度
ax.set_xticks(custom_xticks)
ax.set_xticklabels(custom_xticklabels)
ax.grid(True, which="both", ls="-", alpha=0.2)

plt.tight_layout()
plt.savefig('py_benchmark_acc_rate.png', dpi=output_dpi)
plt.close()

print("Done! Generated: py_benchmark_values.png, py_benchmark_errors.png, py_benchmark_acc_rate.png")